src/bin/pg_autoctl/cli_config.c

```diff
@@ -632,6 +632,157 @@ cli_config_set(int argc, char **argv)
 }
 
 
+static bool
+cli_keeper_config_validate_replication(KeeperConfig *config, PGSQL *pgsql)
+{
+	Keeper keeper = { 0 };
+	NodeAddress primary = { 0 };
+	bool primaryUnreachable;
+
+	if (!keeper_init(&keeper, config))
+	{
+		/* already logged */
+		return false;
+	}
+
+	/* get the primary info to populate the replication source */
+	if (!monitor_get_primary(&keeper.monitor,
+							 config->formation,
+							 keeper.state.current_group,
+							 &primary))
+	{
+		log_warn("Monitor could not be contacted for primary conninfo, skipping "
+				 "password validation");
+		return true;
+	}
+
+	if (primary.nodeId == keeper.state.current_node_id)
+	{
+		/* we are the primary */
+		PGSQL *pgsql = &keeper.postgres.sqlClient;
+
+		/* Skip validation if the primary isn't running. */
+		if (!pg_is_running(config->pgSetup.pg_ctl, config->pgSetup.pgdata))
+		{
+			log_warn("Primary could not be contacted, skipping password "
+					 "validation");
+			return true;
+		}
+
+		/*
+		 * Validate the replication password in a transaction block by setting
+		 * the password using ALTER USER PASSWORD. The transaction block will
+		 * encapsulate the subsequent change to the config file as well and we
+		 * ROLLBACK/COMMIT depending on whether both the ALTER and the config
+		 * file update are successful.
+		 */
+		if (!pgsql_execute(pgsql, "BEGIN"))
+		{
+			log_error("Failed to open a SQL transaction to update the "
+					  "replication password");
+			pgsql_finish(pgsql);
+			return false;
+		}
+
+		if (!pgsql_set_password(pgsql,
+								PG_AUTOCTL_REPLICA_USERNAME,
+								config->replication_password))
+		{
+			log_error("Failed to set replication password, see above for details");
+			pgsql_finish(pgsql);
```

Comment:
Disconnecting without issuing ROLLBACK first might be a good approach, though it needs at least some comments about it.